core/src/syscall/precompiles/blake2s/air.rs

```diff
@@ -0,0 +1,338 @@
+                *local.a[i + 4].prev_value(),
+                local.add2[i].value,
+                local.xor[i + 4],
+                local.shard,
+                &local.channel,
+                local.is_real,
+            );
+            FixedRotateRightOperation::<AB::F>::eval(
+                builder,
+                local.xor[i + 4].value,
+                R_2 as usize,
+                local.rotate_right[i + 4],
+                local.shard,
+                &local.channel,
+                local.is_real,
+            );
+
+            // 2x
+
+            // v[0] = v[0].wrapping_add(v[1]).wrapping_add(m.from_le());
+            Add3Operation::<AB::F>::eval(
+                builder,
+                local.add3[i].value,             // v0 after 1x
+                local.rotate_right[i + 4].value, // v1 after 1x
+                *local.b[i + 4].value(),         // m2
+                local.shard,
+                local.channel,
+                local.is_real,
+                local.add3[i + 4],
+            );
+
+            // v[3] = (v[3] ^ v[0]).rotate_right_const(rd);
+            XorOperation::<AB::F>::eval(
+                builder,
+                local.rotate_right[i].value, // v3 after 1x
+                local.add3[i + 4].value,     // v0 after 1x
+                local.xor[i + 8],
+                local.shard,
+                &local.channel,
+                local.is_real,
+            );
+            FixedRotateRightOperation::<AB::F>::eval(
+                builder,
+                local.xor[i + 8].value,
+                R_3 as usize,
+                local.rotate_right[i + 8],
+                local.shard,
+                &local.channel,
+                local.is_real,
+            );
+
+            // v[2] = v[2].wrapping_add(v[3]);
+            AddOperation::<AB::F>::eval(
+                builder,
+                local.add2[i].value,             // v2 after 1x
+                local.rotate_right[i + 8].value, // v3 after previous operation
+                local.add2[i + 4],
+                local.shard,
+                local.channel,
+                local.is_real.into(),
+            );
+
+            // v[1] = (v[1] ^ v[2]).rotate_right_const(rb);
+            XorOperation::<AB::F>::eval(
+                builder,
+                local.rotate_right[i + 4].value, // v1 after 1x
+                local.add2[i + 4].value,         // v2 after previous operation
+                local.xor[i + 12],
+                local.shard,
+                &local.channel,
+                local.is_real,
+            );
+            FixedRotateRightOperation::<AB::F>::eval(
+                builder,
+                local.xor[i + 12].value,
+                R_4 as usize,
+                local.rotate_right[i + 12],
+                local.shard,
+                &local.channel,
+                local.is_real,
+            );
+        }
+
+        self.constrain_shuffled_indices(builder, &local.shuffled_indices, local.is_real);
```

Comment:
It seems like `local.shuffled_indices` is only used here, and this is just enforcing that the first 4 elements are 0 while the rest are 1.

I believe you can remove the `shuffled_indices` column and its related constraints without any impact. The shuffling is still properly constrained by the constant values of the `vX_shuffle_lookup` vectors offsetting the memory reads.